scipy/stats/_continuous_distns.py

```diff
@@ -3338,6 +3338,220 @@ def _stats(self, mu):
+            vmax = (root2 - m) * np.sqrt(self.pdf(root2, p, b))
+            umax = np.sqrt(self._pdf(m, p, b))
+            rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
+                                     umax=umax, vmin=vmin, vmax=vmax, c=m,
+                                     random_state=rndm)
+        elif b >= min(0.5, 2 * np.sqrt(1 - p) / 3):
+            # ratio of uniforms without mode shift
+            umax = np.sqrt(self._pdf(m, p, b))
+            s = sc.kv(p + 2, b) / sc.kv(p, b)
+            vmax = np.sqrt(self._pdf(self._mode(p + 2, b), p + 2, b) * s)
+            rvs = rvs_ratio_uniforms(lambda x: self.pdf(x, p, b), size=sz,
+                                     umax=umax, vmin=0, vmax=vmax, c=0,
+                                     random_state=rndm)
+        else:
+            # new algorithm in [2]
+            x0 = b / (1 - p)
+            xs = np.max((x0, 2 / b))
+            k1 = self._pdf(m, p, b)
+            A1 = k1 * x0
+            if x0 < 2 / b:
+                k2 = np.exp(-b)
+                if p > 0:
+                    A2 = k2 * ((2 / b)**p - x0**p) / p
+                else:
+                    A2 = k2 * np.log(2 / b**2)
+            else:
+                k2, A2 = 0, 0
+            k3 = xs**(p - 1)
+            A3 = 2 * k3 * np.exp(-xs * b / 2) / b
+            A = A1 + A2 + A3
+
+            size1d = tuple(np.atleast_1d(sz))
+            N = np.prod(size1d)  # number of rvs needed, reshape upon return
+
+            # start sampling
+            rng = self._random_state
+            x = np.zeros(N)
+            simulated = 0
+            # [2]: rejection constant is < 2.73; so expected runtime is finite
+            while True:
+                k = N - simulated
+                h, rvs = np.zeros(k), np.zeros(k)
+                # simulate uniform rvs on [x1, x2] and [0, y2]
+                u1 = rng.random_sample(size=k)
+                v1 = A * rng.random_sample(size=k)
+                cond1 = v1 <= A1
+                cond2 = np.logical_not(cond1) & (v1 <= A1 + A2)
+                cond3 = np.logical_not(cond1 | cond2)
+                # subdomain (0, x0)
+                rvs[cond1] = x0 * v1[cond1] / A1
+                h[cond1] = k1
+                # subdomain (x0, 2 / b)
+                if p > 0:
+                    rvs[cond2] = (x0**p + (v1[cond2] - A1) * p / k2)**(1 / p)
+                else:
+                    rvs[cond2] = b * np.exp((v1[cond2] - A1) * np.exp(b))
+                h[cond2] = k2 * rvs[cond2]**(p - 1)
+                # subdomain (xs, infinity)
+                z = np.exp(-xs * b / 2) - b * (v1[cond3] - A1 - A2) / (2 * k3)
+                rvs[cond3] = -2 / b * np.log(z)
+                h[cond3] = k3 * np.exp(-rvs[cond3] * b / 2)
+                # apply rejection method
+                accept = (u1 * h <= self.pdf(rvs, p, b))
+                num_accept = sum(accept)
+                if num_accept > 0:
```

Comment:
By construction, `num_accept` can't be more than `N - simulated`, so these statements

                if num_accept > 0:
                    take = min(num_accept, N - simulated)
                    x[simulated:(simulated + take)] = rvs[accept][0:take]
                    simulated += take

can be simplified to

                if num_accept > 0:
                    x[simulated:(simulated + num_accept)] = rvs[accept]
                    simulated += num_accept

Reply:
good observation; fixed